src/tscore/DbgCtl.cc

```diff
@@ -0,0 +1,111 @@
+/** @file
+
+  Implementation file for DbgCtl class.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <mutex>
+#include <atomic>
+#include <set>
+#include <cstring>
+
+#include <tscore/Diags.h>
+
+// The resistry of fast debug controllers has a ugly implementation to handle the whole-program initialization
+// order problem with C++.
```

Comment:
Can you explain how this handles the whole-program init order problem? Also, a general summary of the approach here how the debug tags are assembled into this registry and efficiently checked.

Reply:
The registry data is "naturally" just a static member variable.  But the registry may be accessed before initialization for the .cc file defining the registry data.  So the data is instead defined as a static local in a static function that returns a referent as to the data.  This is an ugly but standard solution to the order of initialization problem.

Under the old design, each "Debug()" call stores the debug tag.  When debug was globally enabled, the tag was tested for a match with the configured PCRE for enabled tags.  If it matched, the output was generated.  But, the test for a match required significant computation, causing a big degradation in ATS performance.

Under the new design, each "Debug()" call stores a pointer into entry in the registry set for the tag.  A set entry contains the tag string and an on/off flag for the tag.  So each Debug call only has to dereference the pointer to the set entry, and then check the on/off flag.  When the PCRE for enabled tags is changed, the registry set is iterated over, and  only then is each tag checked for a match to the PCRE.  The on/off flag for the tag is then set based on the check result.